stan/math/prim/fun/quantile.hpp

```diff
@@ -0,0 +1,116 @@
+#ifndef STAN_MATH_PRIM_FUN_QUANTILE_HPP
+#define STAN_MATH_PRIM_FUN_QUANTILE_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/as_array_or_scalar.hpp>
+#include <algorithm>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return sample quantiles corresponding to the given probabilities.
+ * The smallest observation corresponds to a probability of 0 and the largest to
+ * a probability of 1.
+ *
+ * Implements algorithm 7 from Hyndman, R. J. and Fan, Y.,
+ * Sample quantiles in Statistical Packages (R's default quantile function)
+ *
+ * @tparam T Type of input vector
+ * @param xs Numeric vector whose sample quantiles are wanted
+ * @param p Probability with value between 0 and 1.
+ * @return Sample quantile.
+ * @throw std::invalid_argument If any element of xs is NaN, or size 0.
+ * @throw std::domain_error If p<0 or p>1.
+ */
+template <typename T, require_vector_t<T>* = nullptr,
+          require_vector_vt<std::is_arithmetic, T>* = nullptr>
+inline double quantile(const T& xs, const double p) {
+  check_not_nan("quantile", "xs", xs);
+  check_nonzero_size("quantile", "xs", xs);
+
+  check_not_nan("quantile", "p", p);
+  check_bounded("quantile", "p", p, 0, 1);
+
+  size_t n_sample = xs.size();
+
+  Eigen::VectorXd x = as_array_or_scalar(xs);
+
+  if (n_sample == 1)
+    return x.coeff(0);
+  if (p == 0.)
+    return *std::min_element(x.data(), x.data() + n_sample);
+  if (p == 1.)
+    return *std::max_element(x.data(), x.data() + n_sample);
+
+  double index = (n_sample - 1) * p;
+  size_t lo = std::floor(index);
+  size_t hi = std::ceil(index);
+
+  std::sort(x.data(), x.data() + n_sample);
```

Comment:
If you sort once after making `x` can you just do

```
 if (p == 0.)
    return x[0];
  if (p == 1.)
    return x[n_sample];
```

Reply:
Isn't running these checks before sorting more efficient? In the sense that either is true then we save the sorting which is more expensive?

Reply:
Oh good point! Returning immediately so makes sense